File: Hybrid/chaincode/fabcar/javascript/lib/fabcar.js
'use strict';

const { Contract } = require('fabric-contract-api');
const pgp = require('pg-promise')();
const crypto = require('crypto');
const fs = require('fs');
const { exec } = require('child_process');
const path = require('path');
const { env } = require('process');

class FabCar extends Contract {

    // ############################################################################
    // ############################################################################
    //
    //                              UTILS
    //
    // ############################################################################
    // ############################################################################

    async connectToDatabase(timescaleConnStr) {
        return new Promise((resolve, reject) => {
            try {
                const db = pgp(timescaleConnStr);
                console.info('TimescaleDB connection established');
                resolve(db);
            } catch (error) {
                console.error('Error connecting to TimescaleDB:', error);
                reject(error);
            }
        });
    }

    async runQuery(db, query, params = []) {
        return new Promise((resolve, reject) => {
            db.any(query, params)
                .then(results => {
                    resolve(results);
                })
                .catch(error => {
                    console.error('Error executing query:', error);
                    reject(error);
                });
        });
    }

    async createFolderIfNotExists(folderPath) {
        if (!fs.existsSync(folderPath)) {
            fs.mkdirSync(folderPath, { recursive: true });
            console.info(`Created folder: ${folderPath}`);
        }
    }

    async appendToFile(filePath, dataList) {
        if (!fs.existsSync(filePath)) {
            fs.writeFileSync(filePath, '');
            console.info(`Created file: ${filePath}`);
        }

        dataList.forEach(data => {
            const content = `${data.date};${data.data}`;
            fs.appendFileSync(filePath, `\n${content}`);
        });

        console.info(`Appended content to file: ${filePath}`);
    }

    async sleep(ms) {
        return new Promise(resolve => setTimeout(resolve, ms));
    }

    // ############################################################################
    // ############################################################################
    //          
    //                              CHAINCODE
    //
    // ############################################################################
    // ############################################################################

    async initLedger(ctx) {
        console.info('============= START : Initialize Ledger ===========');

        //create database
        const timescaleConnStr = 'postgres://myuser:rootpassword@172.17.0.1:5432/postgres';
        const db = await this.connectToDatabase(timescaleConnStr);

        const dbName = process.env.HOSTNAME;

        // Create the database if it doesn't exist
        await this.runQuery(db, `CREATE DATABASE ${dbName}`);

        // Connect to the new database
        const dbConnStr = `postgres://myuser:rootpassword@172.17.0.1:5432/${dbName}`;
        const dbConnection = await this.connectToDatabase(dbConnStr);

        // Create the _keys table
        await this.runQuery(dbConnection, 'CREATE TABLE IF NOT EXISTS _keys (sensorId TEXT, _key TEXT);');

        console.info('============= END : Initialize Ledger ===========');
    }


    async AddSensor(ctx, sensorstr, a_str, b_str) {
        console.info('============= START : AddSensor ===========');
        const mysqlConnStr = {host: "172.17.0.1", port: 3306, user: "root", password: "rootpassword0", database: process.env.HOSTNAME};
        const connection = await this.connectToDatabase(mysqlConnStr);
        const sensor = JSON.parse(sensorstr);
        const a = BigInt(a_str);
        const b = BigInt(b_str);
        const randomizer = BigInt(7265483); //hardcoded to prevent "leaked a and b during transmission"

        //compute shared key (stored in the bloc and visible for everyone)
        let sk_x = BigInt(BigInt('0x' + Buffer.from(sensor.sharedKey, 'utf-8').toString('hex')));
        let sk_y = randomizer * (a * sk_x + b);
        sensor.sharedKey = sk_x.toString() + '|' + sk_y.toString();

        //compute my key
        let mk_x = BigInt('0x' + crypto.randomBytes(32).toString('hex'));
        let mk_y = randomizer * (a * mk_x + b);
        let mykey = mk_x.toString() + '|' + mk_y.toString();

        //create table for sensor
        await this.runQuery(connection,`CREATE TABLE IF NOT EXISTS ${sensor.sensorId} (date DATETIME DEFAULT CURRENT_TIMESTAMP, data TEXT);`);
        console.info('Created sensor table in MySQL !');

        //store my key into database (other key storing systems can be possible)
        await this.runQuery(connection,'INSERT INTO _keys (sensorId, _key) VALUES (?, ?)',[sensor.sensorId, mykey]);
        console.info('Inserted key into MySQL !');

        //insert bloc
        await ctx.stub.putState(sensor.sensorId, Buffer.from(JSON.stringify(sensor)));
        console.info('============= END : AddSensor ===========');
    }

    async InsertData(ctx, sensorId, datastr) {
        console.info('============= START : InsertData ===========');
        const data = JSON.parse(datastr);
        let results = undefined;

        try {
            const mysqlConnStr = {host: "172.17.0.1", port: 3306, user: "root", password: "rootpassword0", database: process.env.HOSTNAME};
            const connection = await this.connectToDatabase(mysqlConnStr);

            // Get shared Key from blockchain
            const sensorAsBytes = await ctx.stub.getState(sensorId);
            if (!sensorAsBytes || sensorAsBytes.length === 0) {
                throw new Error(`${sensorId} does not exist`);
            }
            let sensorFound = JSON.parse(sensorAsBytes.toString());

            //get my key
            results = await this.runQuery(connection,'SELECT _key FROM _keys WHERE sensorId = ? ',[sensorFound.sensorId]);
            if (results.length == 0) 
                return console.info('No results found for the given sensorId.');
            const mykey = results[0]._key;
            const [sk_x_str, sk_y_str] = mykey.split('|');
            const [mk_x_str, mk_y_str] = sensorFound.sharedKey.split('|');
            const sk_x = BigInt(sk_x_str);
            const sk_y = BigInt(sk_y_str);
            const mk_x = BigInt(mk_x_str);
            const mk_y = BigInt(mk_y_str);

            // Compute secret with Shamir's Secret Sharing
            const a = (sk_y - mk_y) / (sk_x - mk_x);
            const b = mk_y - a * mk_x;
            const secret = a * BigInt(658741143) + b; 

            console.info('Computed secret:', secret.toString());
            // Encrypt data using AES-256 encryption
            const secretBuffer = Buffer.from(secret.toString().slice(-32), 'utf-8'); // Use last 256 bits
            for (const d of data) {
                const iv = Buffer.from('1234567890abcdef1234567890abcdef', 'hex');
                const cipher = crypto.createCipheriv('aes-256-cbc', crypto.createHash('sha256').update(secretBuffer).digest(), iv);
                let encrypted = cipher.update(Buffer.from(d['data']), undefined,'hex');
                encrypted += cipher.final('hex');
                d['data'] = encrypted;
            }

            // Insert data into MySQL
            for (const d of data) 
                await this.runQuery(connection,`INSERT INTO ${mysql.escapeId(sensorId)} (date, data) VALUES (?, ?)`,[d['date'], d['data']]);
        } catch (err) {
            console.info("Failed to add data in MySQL: " + err);
            console.info("Data will be stored to a temporary file!");

            const folderPath = '/tmp/sql_queue';
            const filePath = path.join(folderPath, `${sensorId}.txt`);
            this.createFolderIfNotExists(folderPath);
            this.appendToFile(filePath, data);

        }
        console.info('============= END : InsertData ===========');
    }



    async ReadData(ctx, sensorId) {
        console.info('============= START : ReadData ===========');

        const mysqlConnStr = {host: "172.17.0.1", port: 3306, user: "root", password: "rootpassword0", database: process.env.HOSTNAME};
        const connection = await this.connectToDatabase(mysqlConnStr);
        let results = undefined;
        // Get shared Key from blockchain
        const sensorAsBytes = await ctx.stub.getState(sensorId);
        if (!sensorAsBytes || sensorAsBytes.length === 0) {
            throw new Error(`${sensorId} does not exist`);
        }
        let sensorFound = JSON.parse(sensorAsBytes.toString());


        try{
            results = await this.runQuery(connection,'SELECT _key FROM _keys WHERE sensorId = ? ',[sensorFound.sensorId]);
            console.info('Inserted key into MySQL !');
        }catch{console.error('Error selecting key from MySQL:', error);}

        if (results.length == 0) 
            return console.info('No results found for the given sensorId.');

        const mykey = results[0]._key;
        console.info('Got the shared key from MySQL:', mykey);

        // Compute secret with SSS (Shamir's Secret Sharing)
        const [sk_x_str, sk_y_str] = mykey.split('|');
        const [mk_x_str, mk_y_str] = sensorFound.sharedKey.split('|');

        const sk_x = BigInt(sk_x_str);
        const sk_y = BigInt(sk_y_str);
        const mk_x = BigInt(mk_x_str);
        const mk_y = BigInt(mk_y_str);

        const a = (sk_y - mk_y) / (sk_x - mk_x);
        const b = mk_y - a * mk_x;
        const secret = a * BigInt(658741143) + b; // 658741.145 * 10^6 to handle it as BigInt

        try{
            results = await this.runQuery(connection,`SELECT * FROM ${mysql.escapeId(sensorId)}`);
            console.info('Selected data from MySQL !');
        }catch{console.error('Error selecting data from MySQL:', error);}
        

        const secretBuffer = Buffer.from(secret.toString().slice(-32), 'utf-8'); // Use last 256 bits
        

        for (const d of results) {
            const iv = Buffer.from('1234567890abcdef1234567890abcdef', 'hex');
            const decipher = crypto.createDecipheriv('aes-256-cbc', crypto.createHash('sha256').update(secretBuffer).digest(), iv);
            const encryptedText = d['data'];
            let decrypted = decipher.update(encryptedText, 'hex', 'utf8');
            decrypted += decipher.final('utf8');
            d['data'] = decrypted; // Store the decrypted data
        }
        console.log(results);
        console.info('============= END : ReadData ===========');

        return JSON.stringify(results);
    }
}

module.exports = FabCar;

File: Hybrid/chaincode/fabcar/go/fabcar.go
package main

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/sha256"
	"crypto/rand"
	"database/sql"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"log"
	"math/big"
	"os"
	"strings"
	"time"

	_ "github.com/lib/pq"
	"github.com/hyperledger/fabric-contract-api-go/contractapi"
)

type FabCar struct {
	contractapi.Contract
}

type Sensor struct {
	SensorID  string `json:"sensorId"`
	SharedKey string `json:"sharedKey"`
}

type Data struct {
	Date string `json:"date"`
	Data string `json:"data"`
}

func (s *FabCar) connectToPostgres(dbName string) (*sql.DB, error) {
	var connStr string
	if dbName == "" {
		connStr = "host=172.17.0.1 port=5112 user=postgres password=postgres sslmode=disable"
	} else {
		connStr = fmt.Sprintf("host=172.17.0.1 port=5112 user=postgres password=postgres dbname=%s sslmode=disable", dbName)
	}

	db, err := sql.Open("postgres", connStr)
	if err != nil {
		return nil, fmt.Errorf("failed to open connection: %v", err)
	}

	// Ping the database to check if the connection is actually established
	if err := db.Ping(); err != nil {
		return nil, fmt.Errorf("failed to connect to database: %v", err)
	}

	fmt.Println("Successfully connected to TimescaleDB!")
	return db, nil
}

func (s *FabCar) runQuery(db *sql.DB, query string, args ...interface{}) (*sql.Rows, error) {

	return db.Query(query, args...)
}

func (s *FabCar) createFolderIfNotExists(folderPath string) error {
	return os.MkdirAll(folderPath, os.ModePerm)
}

func (s *FabCar) appendToFile(filePath string, dataList []Data) error {
	f, err := os.OpenFile(filePath, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		fmt.Println(err);return err
	}
	defer f.Close()

	for _, data := range dataList {
		content := fmt.Sprintf("%s;%s\n", data.Date, data.Data)
		if _, err := f.WriteString(content); err != nil {
			fmt.Println(err);return err
		}
	}
	return nil
}

func (s *FabCar) sleep(ms int) {
	time.Sleep(time.Duration(ms) * time.Millisecond)
}

func (s *FabCar) InitLedger(ctx contractapi.TransactionContextInterface) error {
	log.Println("============= START : Initialize Ledger ===========")

	db, err := s.connectToPostgres("")
	if err != nil {
		fmt.Println(err);return err
	}
	defer db.Close()

	dbName := os.Getenv("HOSTNAME")
	fmt.Println(dbName)
	_, err = s.runQuery(db, fmt.Sprintf(`CREATE DATABASE "%s" ;`, dbName))
	if err != nil {
		fmt.Println(err);return err
	}

	db, err = s.connectToPostgres(dbName)
	if err != nil {
		fmt.Println(err);return err
	}
	defer db.Close()

	_, err = s.runQuery(db, "CREATE TABLE IF NOT EXISTS _keys (sensorId TEXT UNIQUE, _key TEXT);")
	if err != nil {
		fmt.Println(err);return err
	}

	log.Println("============= END : Initialize Ledger ===========")
	return nil
}

func (s *FabCar) AddSensorKeys(ctx contractapi.TransactionContextInterface, sensorStr, aStr, bStr string) error {
	log.Println("============= START : AddSensorKeys ===========")

	var sensor Sensor
	err := json.Unmarshal([]byte(sensorStr), &sensor)
	if err != nil {
		fmt.Println(err)
		return err
	}

	a := new(big.Int)
	a.SetString(aStr, 10)

	b := new(big.Int)
	b.SetString(bStr, 10)

	randomizer := big.NewInt(7265483)

	skX := new(big.Int)
	skX.SetString(fmt.Sprintf("%x", sha256.Sum256([]byte(sensor.SharedKey))), 16)

	skY := new(big.Int)
	skY.Mul(a, skX)
	skY.Add(skY, b)
	skY.Mul(skY, randomizer)

	sensor.SharedKey = fmt.Sprintf("%s|%s", skX.String(), skY.String())

	n := 256

	// Générer un grand nombre aléatoire
	mkX, err := rand.Int(rand.Reader, new(big.Int).Lsh(big.NewInt(1), uint(n)))
	if err != nil {
		fmt.Println("Erreur lors de la génération du nombre aléatoire :", err)
		return err
	}


	mkY := new(big.Int)
	mkY.Mul(a, mkX)
	mkY.Add(mkY, b)
	mkY.Mul(mkY, randomizer)

	myKey := fmt.Sprintf("%s|%s", mkX.String(), mkY.String())

	dbName := os.Getenv("HOSTNAME")
	db, err := s.connectToPostgres(dbName)
	if err != nil {
		fmt.Println(err)
		return err
	}
	defer db.Close()

	_, err = s.runQuery(db, fmt.Sprintf("INSERT INTO _keys (sensorId, _key) VALUES ('%s', '%s')", sensor.SensorID, myKey))
	if err != nil {
		fmt.Println(err)
		return err
	}

	sensorJSON, err := json.Marshal(sensor)
	if err != nil {
		fmt.Println(err)
		return err
	}

	err = ctx.GetStub().PutState(sensor.SensorID, sensorJSON)
	if err != nil {
		fmt.Println(err)
		return err
	}

	log.Println("============= END : AddSensorKeys ===========")
	return nil
}

func (s *FabCar) AddSensor(ctx contractapi.TransactionContextInterface, sensorStr, aStr, bStr string) error {
	log.Println("============= START : AddSensor ===========")

	var sensor Sensor
	err := json.Unmarshal([]byte(sensorStr), &sensor)
	if err != nil {
		fmt.Println(err)
		return err
	}

	db, err := s.connectToPostgres("hybrid")
	if err != nil {
		fmt.Println(err)
		return err
	}
	defer db.Close()

	_, err = s.runQuery(db, fmt.Sprintf(`
	CREATE TABLE IF NOT EXISTS %s (
		time TIMESTAMPTZ NOT NULL,
		data TEXT
	);`, sensor.SensorID))
	if err != nil {
		fmt.Println(err)
		return err
	}

	// Convertir la table en hypertable
	_, err = s.runQuery(db, fmt.Sprintf(`
	SELECT create_hypertable('%s', 'time');`, sensor.SensorID))
	if err != nil {
		fmt.Println(err)
		return err
	}

	// Ajouter des index sur la colonne temporelle
	_, err = s.runQuery(db, fmt.Sprintf(`
	CREATE INDEX ON %s (time);`, sensor.SensorID))
	if err != nil {
		fmt.Println(err)
		return err
	}

	log.Println("============= END : AddSensor ===========")
	return nil
}


func (s *FabCar) InsertData(ctx contractapi.TransactionContextInterface, sensorID, dataStr string) error {
	log.Println("============= START : InsertData ===========")
	var data []Data
	err := json.Unmarshal([]byte(dataStr), &data)
	if err != nil {
		fmt.Println(err);return err
	}

	dbName := os.Getenv("HOSTNAME")
	db, err := s.connectToPostgres(dbName)
	if err != nil {
		fmt.Println(err);return err
	}
	defer db.Close()

	sensorAsBytes, err := ctx.GetStub().GetState(sensorID)
	if err != nil {
		fmt.Println(err);return err
	}
	if sensorAsBytes == nil {
		return fmt.Errorf("%s does not exist", sensorID)
	}

	var sensor Sensor
	err = json.Unmarshal(sensorAsBytes, &sensor)
	if err != nil {
		fmt.Println(err);return err
	}

	rows, err := s.runQuery(db, fmt.Sprintf("SELECT _key FROM _keys WHERE sensorId = '%s'", sensor.SensorID))
	if err != nil {
		fmt.Println(err);return err
	}
	defer rows.Close()
	
	var myKey string
	if rows.Next() {
		err := rows.Scan(&myKey)
		if err != nil {
			fmt.Println(err);return err
		}
	} else {
		return fmt.Errorf("no results found for the given sensorId")
	}

	skXStr, skYStr := splitKey(sensor.SharedKey)
	mkXStr, mkYStr := splitKey(myKey)

	skX := new(big.Int)
	skX.SetString(skXStr, 10)

	skY := new(big.Int)
	skY.SetString(skYStr, 10)

	mkX := new(big.Int)
	mkX.SetString(mkXStr, 10)

	mkY := new(big.Int)
	mkY.SetString(mkYStr, 10)

	a := new(big.Int).Sub(skY, mkY)
	a.Div(a, new(big.Int).Sub(skX, mkX))

	b := new(big.Int).Sub(mkY, new(big.Int).Mul(a, mkX))

	secret := new(big.Int).Mul(a, big.NewInt(658741143))
	secret.Add(secret, b)

	secretStr := secret.String()
	fmt.Println("Le secret est ")
	fmt.Println(secretStr)

	if len(secretStr) < 32 {
		secretStr = fmt.Sprintf("%032s", secretStr)
	}
	secretBuffer := sha256.Sum256([]byte(secretStr[len(secretStr)-32:]))


	db, err = s.connectToPostgres("hybrid")
	if err != nil {
		fmt.Println(err);return err
	}
	defer db.Close()

	// Construire la requête d'insertion
	query := fmt.Sprintf("INSERT INTO %s (time, data) VALUES ", sensorID)

	var values []string
	for _, d := range data {
		ciphertext, err := encrypt(secretBuffer[:], []byte(d.Data))
		if err != nil {
			fmt.Println(err)
			return err
		}
		encryptedData := hex.EncodeToString(ciphertext)
		value := fmt.Sprintf("('%s', '%s')", d.Date, encryptedData)
		values = append(values, value)
	}

	// Ajouter toutes les valeurs à la requête
	query += strings.Join(values, ", ")

	fmt.Println("Executing INSERT query ... ")
	// Exécuter la requête d'insertion
	_, err = s.runQuery(db, query)
	if err != nil {
		fmt.Println(err)
		return err
	}


	log.Println("============= END : InsertData ===========")
	return nil
}




func (s *FabCar) ReadData(ctx contractapi.TransactionContextInterface, sensorID string) (string, error) {
	log.Println("============= START : ReadData ===========")
	dbName := os.Getenv("HOSTNAME")
	db, err := s.connectToPostgres(dbName)
	if err != nil {
		return "",err
	}
	defer db.Close()


	sensorAsBytes, err := ctx.GetStub().GetState(sensorID)
	if err != nil {
		return "", err
	}
	if sensorAsBytes == nil {
		return "", fmt.Errorf("%s does not exist", sensorID)
	}

	var sensor Sensor
	err = json.Unmarshal(sensorAsBytes, &sensor)
	if err != nil {
		return "", err
	}

	rows, err := s.runQuery(db, fmt.Sprintf("SELECT _key FROM _keys WHERE sensorId = '%s'", sensor.SensorID))
	if err != nil {
		return "", err
	}
	defer rows.Close()

	var myKey string
	if rows.Next() {
		err := rows.Scan(&myKey)
		if err != nil {
			return "", err
		}
	} else {
		return "", fmt.Errorf("no results found for the given sensorId")
	}

	skXStr, skYStr := splitKey(myKey)
	mkXStr, mkYStr := splitKey(sensor.SharedKey)

	skX := new(big.Int)
	skX.SetString(skXStr, 10)

	skY := new(big.Int)
	skY.SetString(skYStr, 10)

	mkX := new(big.Int)
	mkX.SetString(mkXStr, 10)

	mkY := new(big.Int)
	mkY.SetString(mkYStr, 10)

	a := new(big.Int).Sub(skY, mkY)
	a.Div(a, new(big.Int).Sub(skX, mkX))

	b := new(big.Int).Sub(mkY, new(big.Int).Mul(a, mkX))

	secret := new(big.Int).Mul(a, big.NewInt(658741143))
	secret.Add(secret, b)

	secretStr := secret.String()
	if len(secretStr) < 32 {
		secretStr = fmt.Sprintf("%032s", secretStr)
	}
	secretBuffer := sha256.Sum256([]byte(secretStr[len(secretStr)-32:]))


	db, err = s.connectToPostgres("hybrid")
	if err != nil {
		return "",err
	}
	defer db.Close()


	rows, err = s.runQuery(db, fmt.Sprintf("SELECT * FROM %s", sensorID))
	if err != nil {
		return "", err
	}
	defer rows.Close()

	var results []Data
	for rows.Next() {
		var d Data
		err := rows.Scan(&d.Date, &d.Data)
		if err != nil {
			return "", err
		}

		plaintext, err := decrypt(secretBuffer[:], d.Data)
		if err != nil {
			return "", err
		}
		d.Data = string(plaintext)

		results = append(results, d)
	}

	resultJSON, err := json.Marshal(results)
	if err != nil {
		return "", err
	}

	log.Println("============= END : ReadData ===========")
	return string(resultJSON), nil
}

func (s *FabCar) SpeedTestInsert(ctx contractapi.TransactionContextInterface, sensorID string) error {
	log.Println("============= START : InsertEncryptedData ===========")
	dbName := os.Getenv("HOSTNAME")
	db, err := s.connectToPostgres(dbName)
	if err != nil {
		return err
	}
	defer db.Close()

	sensorAsBytes, err := ctx.GetStub().GetState(sensorID)
	if err != nil {
		return err
	}
	if sensorAsBytes == nil {
		return fmt.Errorf("%s does not exist", sensorID)
	}

	var sensor Sensor
	err = json.Unmarshal(sensorAsBytes, &sensor)
	if err != nil {
		return err
	}

	rows, err := s.runQuery(db, fmt.Sprintf("SELECT _key FROM _keys WHERE sensorId = '%s'", sensor.SensorID))
	if err != nil {
		return err
	}
	defer rows.Close()

	var myKey string
	if rows.Next() {
		err := rows.Scan(&myKey)
		if err != nil {
			return err
		}
	} else {
		return fmt.Errorf("no results found for the given sensorId")
	}

	skXStr, skYStr := splitKey(myKey)
	mkXStr, mkYStr := splitKey(sensor.SharedKey)

	skX := new(big.Int)
	skX.SetString(skXStr, 10)

	skY := new(big.Int)
	skY.SetString(skYStr, 10)

	mkX := new(big.Int)
	mkX.SetString(mkXStr, 10)

	mkY := new(big.Int)
	mkY.SetString(mkYStr, 10)

	a := new(big.Int).Sub(skY, mkY)
	a.Div(a, new(big.Int).Sub(skX, mkX))

	b := new(big.Int).Sub(mkY, new(big.Int).Mul(a, mkX))

	secret := new(big.Int).Mul(a, big.NewInt(658741143))
	secret.Add(secret, b)

	secretStr := secret.String()
	if len(secretStr) < 32 {
		secretStr = fmt.Sprintf("%032s", secretStr)
	}
	secretBuffer := sha256.Sum256([]byte(secretStr[len(secretStr)-32:]))

	db, err = s.connectToPostgres("hybrid")
	if err != nil {
		return err
	}
	defer db.Close()

	startTime := time.Date(2023, 1, 1, 0, 0, 0, 0, time.UTC)
	endTime := time.Date(2023, 1, 9, 0, 0, 0, 0, time.UTC)
	totalData := int(endTime.Sub(startTime).Seconds())
	data := make([]Data, totalData)

	for i := 0; i < totalData; i++ {
		data[i] = Data{
			Date: startTime.Add(time.Duration(i) * time.Second).Format(time.RFC3339),
			Data: fmt.Sprintf("%d",i),
		}
	}

	batchSize := 50000
	for i := 0; i < totalData; i += batchSize {
		end := i + batchSize
		if end > totalData {
			end = totalData
		}

		query := fmt.Sprintf("INSERT INTO %s (time, data) VALUES ", sensorID)
		var values []string

		for _, d := range data[i:end] {
			ciphertext, err := encrypt(secretBuffer[:], []byte(d.Data))
			if err != nil {
				fmt.Println(err)
				return err
			}
			encryptedData := hex.EncodeToString(ciphertext)
			value := fmt.Sprintf("('%s', '%s')", d.Date, encryptedData)
			values = append(values, value)
		}

		query += strings.Join(values, ", ")

		fmt.Println("Executing INSERT query ... ")
		_, err = s.runQuery(db, query)
		if err != nil {
			fmt.Println(err)
			return err
		}

		log.Printf("Inserted %d/%d data points\n", end, totalData)
		s.sleep(1000)
	}

	log.Println("============= END : InsertEncryptedData ===========")
	return nil
}



func (s *FabCar) SpeedTestQuery(ctx contractapi.TransactionContextInterface, sensorID string) (string, error) {
	log.Println("============= START : speedTest ===========")
	dbName := os.Getenv("HOSTNAME")
	db, err := s.connectToPostgres(dbName)
	if err != nil {
		return "",err
	}
	defer db.Close()


	sensorAsBytes, err := ctx.GetStub().GetState(sensorID)
	if err != nil {
		return "", err
	}
	if sensorAsBytes == nil {
		return "", fmt.Errorf("%s does not exist", sensorID)
	}

	var sensor Sensor
	err = json.Unmarshal(sensorAsBytes, &sensor)
	if err != nil {
		return "", err
	}

	rows, err := s.runQuery(db, fmt.Sprintf("SELECT _key FROM _keys WHERE sensorId = '%s'", sensor.SensorID))
	if err != nil {
		return "", err
	}
	defer rows.Close()

	var myKey string
	if rows.Next() {
		err := rows.Scan(&myKey)
		if err != nil {
			return "", err
		}
	} else {
		return "", fmt.Errorf("no results found for the given sensorId")
	}

	skXStr, skYStr := splitKey(myKey)
	mkXStr, mkYStr := splitKey(sensor.SharedKey)

	skX := new(big.Int)
	skX.SetString(skXStr, 10)

	skY := new(big.Int)
	skY.SetString(skYStr, 10)

	mkX := new(big.Int)
	mkX.SetString(mkXStr, 10)

	mkY := new(big.Int)
	mkY.SetString(mkYStr, 10)

	a := new(big.Int).Sub(skY, mkY)
	a.Div(a, new(big.Int).Sub(skX, mkX))

	b := new(big.Int).Sub(mkY, new(big.Int).Mul(a, mkX))

	secret := new(big.Int).Mul(a, big.NewInt(658741143))
	secret.Add(secret, b)

	secretStr := secret.String()
	if len(secretStr) < 32 {
		secretStr = fmt.Sprintf("%032s", secretStr)
	}
	secretBuffer := sha256.Sum256([]byte(secretStr[len(secretStr)-32:]))


	db, err = s.connectToPostgres("hybrid")
	if err != nil {
		return "",err
	}
	defer db.Close()


	query := fmt.Sprintf(`
		SELECT time_bucket('1 day', time) AS day, data 
		FROM %s 
		WHERE time >= '2023-01-01' AND time < '2023-01-02';`, sensorID)

	rows, err = s.runQuery(db, query)
	if err != nil {
		return "",err
	}
	defer rows.Close()


	var results []Data
	for rows.Next() {
		var d Data
		err := rows.Scan(&d.Date, &d.Data)
		if err != nil {
			return "", err
		}

		plaintext, err := decrypt(secretBuffer[:], d.Data)
		if err != nil {
			return "", err
		}
		d.Data = string(plaintext)

		results = append(results, d)
	}

	resultJSON, err := json.Marshal(results)
	if err != nil {
		return "", err
	}
	
	log.Println("============= END : speedTest ===========")
	return string(resultJSON), nil
}


func splitKey(key string) (string, string) {
	parts := strings.Split(key, "|")
	return parts[0], parts[1]
}

func encrypt(key []byte, plaintext []byte) ([]byte, error) {
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, err
	}

	iv := []byte("1234567890abcdef")
	cfb := cipher.NewCFBEncrypter(block, iv)
	ciphertext := make([]byte, len(plaintext))
	cfb.XORKeyStream(ciphertext, plaintext)
	return ciphertext, nil
}

func decrypt(key []byte, ciphertext string) ([]byte, error) {
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, err
	}

	iv := []byte("1234567890abcdef")
	ciphertextBytes, _ := hex.DecodeString(ciphertext)
	cfb := cipher.NewCFBDecrypter(block, iv)
	plaintext := make([]byte, len(ciphertextBytes))
	cfb.XORKeyStream(plaintext, ciphertextBytes)
	return plaintext, nil
}


func main() {
	chaincode, err := contractapi.NewChaincode(new(FabCar))
	if err != nil {
		log.Fatalf("Error creating FabCar chaincode: %s", err)
	}

	if err := chaincode.Start(); err != nil {
		log.Fatalf("Error starting FabCar chaincode: %s", err)
	}
}
